cpp/src/arrow/util/functional.h

```diff
@@ -0,0 +1,59 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <tuple>
+
+#include "arrow/util/macros.h"
+
+namespace arrow {
+
+/// Helper struct for examining lambdas and other callables.
+/// If the callable is not overloaded, the argument types of its call operator can be
+/// extracted via single_call::argument_type<Index, Function>
+struct single_call {
+  template <typename R, typename... A>
+  static std::true_type check_impl(R(A...));
```

Comment:
Is this an internal helper? If so, make it private.

Reply:
okay